### PR TITLE
MultiDiscrete to discrete gym action space wrapper

### PR DESCRIPTION
**Pull request by BFAnas**
- [ ] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [ ] algorithm implementation fix
    + [ ] documentation modification
    + [X] new feature
- [X] I have reformatted the code using `make format` (**required**)
- [X] I have checked the code using `make commit-checks` (**required**)
- [X] If applicable, I have mentioned the relevant/related issue(s)
- [ ] If applicable, I have listed every items in this Pull Request below

Has been tested to work with DQN and a custom MultiDiscrete gym env. #643 #311 #236

### CODE REVIEW

File: tianshou/env/gym_wrappers.py [hunk 1]
```diff
@@ -26,3 +26,28 @@ def __init__(self, env: gym.Env, action_per_branch: int) -> None:
...
+
+    def __init__(self, env: gym.Env) -> None:
+        super().__init__(env)
+        assert isinstance(env.action_space, gym.spaces.MultiDiscrete)
+        self.num_dim = env.action_space.shape[0]
+        assert len(
+            set(env.action_space.nvec)
+        ) == 1  # TODO support for different num of actions per dim
+        self.action_per_dim = env.action_space.nvec[0]
+        self.action_space = gym.spaces.Discrete(self.action_per_dim**self.num_dim)
```
Comment (nuance1979) [inline tianshou/env/gym_wrappers.py]: I would suggest supporting different numbers of actions per dim like this:

```python
nvec = env.action_space.nvec
assert nvec.ndim == 1
self.bases = np.ones_like(nvec)
for i in range(1, len(self.bases)):
  self.bases[i] = self.bases[i-1]*nvec[-i]
self.action_space = gym.spaces.Discrete(np.prod(nvec))
```

File: tianshou/env/gym_wrappers.py [hunk 2]
```diff
@@ -26,3 +26,28 @@ def __init__(self, env: gym.Env, action_per_branch: int) -> None:
     def action(self, act: np.ndarray) -> np.ndarray:
         # modify act
         return np.array([self.mesh[i][a] for i, a in enumerate(act)])
+
+
+class MultiDiscreteToDiscrete(gym.ActionWrapper):
+    """Gym environment wrapper to discrete action in multidiscrete environment.
+
+    :param gym.Env env: gym environment with continuous action space.
```
Comment (nuance1979) [inline tianshou/env/gym_wrappers.py]: continuous -> multi-discrete

File: tianshou/env/gym_wrappers.py [hunk 3]
```diff
@@ -26,3 +26,28 @@ def __init__(self, env: gym.Env, action_per_branch: int) -> None:
...
+        self.action_per_dim = env.action_space.nvec[0]
+        self.action_space = gym.spaces.Discrete(self.action_per_dim**self.num_dim)
+
+    def action(self, act: np.ndarray) -> np.ndarray:
+        # modify act
+        converted_act = []
+        for i in range(self.num_dim):
+            converted_act.append(act // self.action_per_dim**(self.num_dim - i))
+            act = act % self.action_per_dim**(self.num_dim - i)
+        return np.array(converted_act)
```
Comment (nuance1979) [inline tianshou/env/gym_wrappers.py]:
```python
converted_act = []
for b in reverse(self.bases):
    converted_act.append(act // b)
    act = act % b
return np.array(converted_act).transpose()
```

Comment (nuance1979) [inline tianshou/env/gym_wrappers.py:53]:
```python
converted_act = []
for b in reverse(self.bases):
    converted_act.append(act // b)
    act = act % b
return np.array(converted_act).transpose()
```

Comment (BFAnas) [reply tianshou/env/gym_wrappers.py]: Hi, thanks for your comments, would you define self.bases like this:
self.bases = [b**(self.num_dim - i) for i, b in enumerate(env.action_space.nvec)]

Comment (nuance1979) [reply tianshou/env/gym_wrappers.py]: Sorry for some reason my previous comment wasn't showing up. Please see my other comment.

Comment (BFAnas) [reply tianshou/env/gym_wrappers.py:53]: `converted_act.append(act // b)` act is an array, so this is not correct. Can you review and correct it?

**Comment by nuance1979** [comment]
Can you please add a test in `test/base/test_env.py`? It should check the conversion works correctly for both a single action and a batch of actions. See examples in the same file.

**Review by Trinkle23897 — DISMISSED** [review]

**Review by Trinkle23897 — APPROVED** [review]

**Pull request merged**